learn/json_preprocessor.go

```diff
@@ -0,0 +1,54 @@
+package learn
+
+import (
+	"io"
+	"unicode"
+)
+
+type stripControlCharactersReader struct {
+	wrapped io.Reader
+}
+
+func newStripControlCharactersReader(wrapped io.Reader) stripControlCharactersReader {
+	return stripControlCharactersReader{wrapped: wrapped}
+}
+
+// Read up to len(p) bytes, removing any unescaped control characters found.
+// Removed characters do not count toward the total bytes read.  Returns the
+// number of bytes read.
+func (r stripControlCharactersReader) Read(p []byte) (n int, err error) {
+	pIdx := 0
+
+	// Read up to len(p) bytes, then discard any control characters.  Continue
+	// reading (and discarding control characters) until p is full or there are
+	// no more bytes to read.
+	for pIdx < len(p) {
+		remaining := len(p) - pIdx
+		buf := make([]byte, remaining)
+
+		var bufN int
+		bufN, err = r.wrapped.Read(buf)
+
+		// Copy from buf to p, skipping unescaped control characters.
+		hasPrecedingSlash := false
+		for _, r := range string(buf) {
+			if unicode.IsControl(r) && !hasPrecedingSlash {
+				continue
+			}
```

Comment:
I'm a bit puzzled by this, the sequence '\' , '\n' is not valid either.  I don't think we need to pay attention to \ to make this work.

If there's a case I'm not getting, can you put a test case in for it?

Reply:
I think the idea is to avoid turning the closed string `"\<CR>"` into the open string `"\"`.

Reply:
Come to think of it, instead of preserving escaped control characters, it may be better to discard both the control character and its escaping backslash, so that `"\<CR>"` turns into the empty string instead of being kept.

Reply:
Hm, good point.  We want to avoid removing escaped control characters, e.g. `"foo\\r"` should be left alone, but I suppose that's an issue when encoding rather than decoding -- by the time we parse it, the rune is either a control character or not.

I [added a field](https://github.com/akitasoftware/akita-cli/blob/0ed1dd0a60f7c5b66fce570a5f0475ee721b1e24/learn/parse_http_test.go#L68) with an escaped CRLF to the JSON object we use in the unit tests, but the test passes after removing the `hasPrecedingSlash` logic above.

Fixed in fea612b431db1793c20b496fc9c34c66556efafc.

Reply:
I think you wanted `"foo\\\r"` instead of `"foo\\r"`.

Reply:
Sorry @liujed, didn't see your comment until I refreshed the page.

I was mistaken -- the JSON parser doesn't support escaping control characters in strings.  E.g. `"foo\\\r"` fails to parse.  You can write `"foo\\r"`, which escapes the slash and avoids the control character entirely.